src/nvim/api/buffer.c

```diff
@@ -402,7 +431,7 @@ void nvim_buf_set_lines(uint64_t channel_id,
     mark_adjust((linenr_T)start, (linenr_T)(end - 1), MAXLNUM, extra, false);
   }
 
-  changed_lines((linenr_T)start, 0, (linenr_T)end, (long)extra);
+  changed_lines((linenr_T)start, 0, (linenr_T)end, (long)extra, true);
```

Comment:
Wait, does a call to `changed_lines` in `buf_set_lines` mean I'll get all of my own changes reflected back at me? Can I opt out of this somehow, at least for my own client's calls to `buf_set_lines`? This will definitely slow down my vim embed.

Reply:
@lunixbochs: Yes you'll get all your changes back, but I can't think of a simple way to make neovim aware which calls from `buf_set_lines()` are coming from your plugin so that it can skip just those ones.

Ultimately it's going to result in double the amount of RPC calls as neovim reflects everything back at you, and while "double the amount" seems like a lot, keep in mind that the biggest advantage of live updates isn't even related to the quantity of RPC traffic - the big advantage is that plugins don't need to stop and re-parse the entire buffer contents when only one line or character has changed.

Reply:
It's far worse than you imagine. It's twice the latency, and amplifies O(N) where N can be and often is HUGE. Please don't downplay this, I know my use case and this is extremely bad for my use case.

Reply:
@lunixbochs the notification must not be silently dropped, the sequence of notifications must form a "single source of truth" from the nvim side. But we could avoid the text data being sent being echoed back, by adding   an extension where client can provide its own update-id to set_lines, then a further notify _might_ reference this update instead of sending lines back (  "this change was done in this place in the stream of diffs") but it also _might  not_, if an error stopped the call from being done (or errored out itself), or the change got coalesced with another change, so the exact diff didn't become part of the stream. 

But if the usecase only is "completely resetting the buffer in exceptional cases", I would expect it to be good enough to call_atomic with three requests, where you turn off updates, do the set_lines overwriting the entire buffer, and turn updates on again. (with the "no initial update" flag that needs to be added, of course).

Reply:
Also, are you sure utf-8 conversion is needed? Is ST3 a UTF-32 editor, or is it just that the python interface to ST3 and/or neovim by _default_ uses `unicode` string rather than `bytes` ? The neovim python-client at least can be configured to use "bytes all the way"

Reply:
> the sequence of notifications must form a "single source of truth" from the nvim side

I vehemently disagree, because in this case *my side* is a valid source of truth. My use case is made vastly simpler by not designing for stuff like embedded terminals and collab plugins from the vim side which will perform unexpected async transformations on the buffer. I understand the tradeoffs and make these to completely prevent UI flicker and reduce UI latency (I believe the ActualVim latency is currently somewhere between 3%-16% that of Oni, so I feel pretty vindicated).

> if the usecase only is "completely resetting the buffer in exceptional cases"

This is by no means an "exceptional case" any more than all of ActualVim is an exceptional case. It does both full and sub-buffer updates from the Sublime side that can happen more than once a second. Toggling LiveUpdate or using placeholder lines don't feel like an acceptable or complete solution because those still cause extra RPC messages when I'm racing latency on the UI thread.

I think I will literally never want my `nvim_buf_set_lines` calls to receive a LiveUpdate notification of any kind, and still definitely not containing the line contents. I also think it's not helpful to use a placeholder identifier, because the other plugin channels would know nothing about its contents.

Reply:
> Also, are you sure utf-8 conversion is needed?
Yes.

Reply:
> I vehemently disagree, because in this case my side is a valid source of truth.

I just meant how the nvim API should be designed: A single uninterrupted LiveUpdate stream should always give a complete conistent stream of the buffer changes. Then of course actualvim can do whatever, if you know actualvim is the only source of changes apart from core nvim then your approach should work perfectly fine. But the api preferably should be defined in terms of the fully general situation (multiple async sources), and then _optimized_  if need be to also work well for your case.

> This is by no means an "exceptional case" any more than all of ActualVim is an exceptional case.

I meant "exceptional" not in judgmental way nor as as statement of relative frequency, just from the consequences: that the previous buffer state could be completely ignored. Only in this "exceptional" case the `call_atomic` workaround I suggested would be safe.

> I think I will literally never want my nvim_buf_set_lines calls to receive a LiveUpdate notification of any kind, and still definitely not containing the line contents.

Right, but if you don't care about it, you do not need to wait for it, and can just silently drop this small notification (it should almost always be "placeholder" with your style of interaction I think) if/when it comes. A notification not waited for is not a "roundtrip", and shouldn't contribute much to latency.

>  I also think it's not helpful to use a placeholder identifier, because the other plugin channels would know nothing about its contents.

Of course only the source channel will receive the placeholder, other subscribers will receive full contents.

Reply:
> I could special case to drop the update message before the python-client finished decoding it,

I was not talking of special casing, rather the entire api can be configured to use `bytes` by supplying `decode=False` to `neovim.attach`. But if the ST3 API forces one to use `str` rather than `bytes`, then I guess it does not help much...

Reply:
Sounds okay until a benchmark proves otherwise. I'd probably use 0 for my placeholder.